File: rust/src/config/args.rs
use clap::Parser;
use std::collections::HashMap;
use std::path::Path;

#[derive(Parser, Debug)]
#[command(author, version, about = "Microservice Toolbox CLI Parser", long_about = None)]
pub struct RawArgs {
    #[arg(long)]
    pub name: Option<String>,

    #[arg(long)]
    pub host: Option<String>,

    #[arg(long)]
    pub port: Option<u16>,

    #[arg(long)]
    pub grpc_host: Option<String>,

    #[arg(long)]
    pub grpc_port: Option<u16>,

    #[arg(long)]
    pub conf: Option<String>,

    #[arg(long)]
    pub log_level: Option<String>,

    #[arg(long)]
    pub key: Option<String>,

    /// Specific arguments in KEY=VALUE format
    #[arg(short, long)]
    pub extra: Vec<String>,
}

#[derive(Debug, Default)]
/// ToolboxArgs provides a standardized CLI interface for microservices.
/// 
/// Security & Reliability (Docker Guard):
/// If DOCKER_ENV=true, settings for host, port, grpc_host, and grpc_port 
/// are strictly IGNORED. This prevents brittle hardcoded overrides from breaking 
/// internal network-aware resolution in dynamic container environments.
pub struct ToolboxArgs {
    pub name: Option<String>,
    pub host: Option<String>,
    pub port: Option<u16>,
    pub grpc_host: Option<String>,
    pub grpc_port: Option<u16>,
    pub conf: Option<String>,
    pub log_level: Option<String>,
    pub key: Option<String>,
    pub extras: HashMap<String, String>,
}

impl ToolboxArgs {
    pub fn parse_cli_args() -> Self {
        let raw = RawArgs::parse();
        let mut result = ToolboxArgs::default();

        // Docker Guard
        let is_docker = Path::new("/.dockerenv").exists()
            || std::env::var("DOCKER_ENV").is_ok_and(|v| v == "true");

        result.name = raw.name;
        result.conf = raw.conf;
        result.log_level = raw.log_level;
        result.key = raw.key;

        // If key provided, set it as ENV override for the decryption engine
        if let Some(k) = &result.key {
            unsafe {
                std::env::set_var("BASTIEN_PRIVATE_KEY_PATH", k);
            }
        }

        if is_docker {
            if raw.host.is_some() || raw.port.is_some() || raw.grpc_host.is_some() || raw.grpc_port.is_some() {
                println!("Toolbox (Rust): Running in Docker. Ignoring CLI network overrides to preserve network-aware resolution.");
            }
            result.host = None;
            result.port = None;
            result.grpc_host = None;
            result.grpc_port = None;
        } else {
            result.host = raw.host;
            result.port = raw.port;
            result.grpc_host = raw.grpc_host;
            result.grpc_port = raw.grpc_port;
        }

        for item in raw.extra {
            if let Some((k, v)) = item.split_once('=') {
                result.extras.insert(k.to_string(), v.to_string());
            }
        }

        result
    }
}

File: go/pkg/config/args.go
package config

import (
	"fmt"
	"os"

	"github.com/spf13/pflag"
)

// CLIArgs holds the standard microservice arguments.
type CLIArgs struct {
	Name     string
	Host     string
	Port     int
	GRPCHost string
	GRPCPort int
	Conf     string
	LogLevel string
	Key      string
	Extra    map[string]string
}

// ParseCLIArgs parses standard and specific command line arguments.
// It implements a "Docker Guard": if DOCKER_ENV=true, --host and --port are ignored.
func (ac *AppConfig) ParseCLIArgs(specificFlags []string) *CLIArgs {
	fs := pflag.NewFlagSet(os.Args[0], pflag.ContinueOnError)

	// Standard flags
	name := fs.String("name", "", "Service name")
	host := fs.String("host", "", "Binding host IP")
	port := fs.Int("port", 0, "Binding port")
	grpcHost := fs.String("grpc_host", "", "GRPC Binding host IP")
	grpcPort := fs.Int("grpc_port", 0, "GRPC Binding port")
	conf := fs.String("conf", "", "Path to configuration file")
	logLevel := fs.String("log_level", "", "Logging level (DEBUG, INFO, etc.)")
	key := fs.String("key", "", "Path to RSA Public/Private key")

	// Dynamic flags for extra arguments
	extras := make(map[string]*string)
	for _, f := range specificFlags {
		extras[f] = fs.String(f, "", fmt.Sprintf("Specific flag: %s", f))
	}

	// Parse
	if err := fs.Parse(os.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			os.Exit(0)
		}
		ac.Logger.Error("Error parsing flags: %v", err)
	}

	result := &CLIArgs{
		Name:     *name,
		Conf:     *conf,
		LogLevel: *logLevel,
		Key:      *key,
		Extra:    make(map[string]string),
	}

	// If key provided, set it as ENV override for the decryption engine (Private Key)
	if *key != "" {
		os.Setenv("BASTIEN_PRIVATE_KEY_PATH", *key)
	}

	// Apply Docker Guard
	if ac.Resolver.IsDocker {
		if *host != "" || *port != 0 || *grpcHost != "" || *grpcPort != 0 {
			ac.Logger.Info("Toolbox: Running in Docker. Ignoring CLI overrides for network flags to preserve network-aware resolution.")
		}
		result.Host = ""
		result.Port = 0
		result.GRPCHost = ""
		result.GRPCPort = 0
	} else {
		result.Host = *host
		result.Port = *port
		result.GRPCHost = *grpcHost
		result.GRPCPort = *grpcPort
	}

	result.Conf = *conf
	for k, v := range extras {
		result.Extra[k] = *v
	}

	return result
}
